### PR TITLE
#37 visualize contacts

### LINKED ISSUE #37: Visual contacts
Visualize contacts in the contact entities; include such data as penetration depth and number of contacts. This is important for verifying contact implementations

### PR DESCRIPTION
**Pull request by nmm0**

### CODE REVIEW

File: src/nimble_contact_entity.h [hunk 1]
```diff
@@ -436,6 +446,8 @@ namespace nimble {
     //   next 3 bits are the face ordinal (range is 1-6)
     //   remaining bits are the genesis element id from the parent FEM mesh (e.g., the global id of the hex from which the face was extracted)
     int contact_entity_global_id_ = -1;
+
+    bool contact_status_ = false;
```
Comment (nmm0) [inline src/nimble_contact_entity.h:450]: Part of the changes in this PR add a contact_status_ member to ContactEntity which indicates the last iteration status. This is mostly used for visualization, but could be used for other purposes

File: src/contact/parallel/bvh_contact_manager.cc [hunk 2]
```diff
@@ -58,6 +58,20 @@ namespace nimble {
...
+      face.set_contact_status( std::count_if( m_last_results.begin(), m_last_results.end(),
+                          [&face]( NarrowphaseResult &res ){ return res.second_global_id == face.contact_entity_global_id(); } ) > 0 );
+    }
+
+    for ( auto &&node : contact_nodes_ )
+    {
+      node.set_contact_status( std::count_if( m_last_results.begin(), m_last_results.end(),
+                                              [&node]( NarrowphaseResult &res ){ return res.first_global_id == node.contact_entity_global_id(); } ) > 0 );
+    }
+
```
Comment (uhetmaniuk) [inline src/contact/parallel/bvh_contact_manager.cc:74]: @nmm0 
Is the purpose of these 2 loops to extract the faces and nodes in contact?
So the quantities calculated by ArborX in the pair (indices, offset)?

**Review by uhetmaniuk — APPROVED** [review]

File: src/contact/parallel/bvh_contact_manager.cc (hunk 2 above)
Comment (nmm0) [reply src/contact/parallel/bvh_contact_manager.cc:74]: Almost; contact_status_ is just a boolean used for debugging so it doesn't currently represent what other elements it is in collision with. It just states that the particular ContactEntity is involved in a collision. I don't think there is a good way to represent in the exodus file what pairs of collisions there are

Comment (rjones-sandia) [reply src/contact/parallel/bvh_contact_manager.cc:74]: What about use the face id for the node "status"? 0 for no collision. The compliment doesn't work i.e. faces can have many node contacts. Here we could count the number of nodes that contact the face but this info is superfluous

**Pull request merged**